### PR TITLE
Post Editor: Always force iframe in pattern editor

### PR DESCRIPTION
**Pull request by t-hamano**
Noticed this while reviewing #65885

## What?

In the post editor, force the pattern editor to always be an iframe, regardless of whether the main editor is an ifarme or not. This will make the pattern editor's height fit its content.

| Before | After |
|--------|--------|
| image| image | 

## Why?

Whether the pattern editor in the post editor is an iframe or not is determined by whether the main editor is an iframe or not.

However, the pattern editor is a resizable editor and the editor height is expected to fit the content. Additionally, the padding appender (the functionality that adds a new block when you click below the last block) is not expected to be used with the pattern editor.

Therefore, if the main editor is not an iframe, clicking below the content in the pattern editor does nothing.

## How?

Add a post type check.

However, changing the conditions that determine if it should be an iframe in a lower component (`VisualEditor`) may not be ideal - whether it should be an iframe or not is something the post editor should only care about. In the future, it may be possible to aggregate these conditions at a higher layer ([`useShouldIframe()` hook](https://github.com/WordPress/gutenberg/blob/9d67345a00a0ce3f1ca40b5894e45fc3b9088857/packages/edit-post/src/components/layout/use-should-iframe.js#L32-L37)).

**Update:** I consolidated all the conditions that determine whether to run an iframe or not into the useShouldIframe hook.

## Testing Instructions

- To run the editor as a non-iframe, you need to set it up as follows:
  - Enable a classic theme.
  - Register a block whose API version is less than or equal to 2. For example, enable [the CoBlocks plugin](https://wordpress.org/plugins/coblocks/).
- Open the post editor.
- Make sure the editor is NOT an iframe.
- Create a synced pattern and click the "Edit original" button.
- Make sure the pattern editor is an iframe.

### CODE REVIEW

**Comment by stokesman** [comment]
I agree this should be fixed. It’s great you spotted the bug.

> changing the conditions that determine if it should be an iframe in a lower component (`VisualEditor`) may not be ideal
…
> In the future, it may be possible to aggregate these conditions at a higher layer ([`useShouldIframe()` hook](https://github.com/WordPress/gutenberg/blob/9d67345a00a0ce3f1ca40b5894e45fc3b9088857/packages/edit-post/src/components/layout/use-should-iframe.js#L32-L37)).

It’s not clear to me why the conditions can’t already be aggregated there. The hook already checks the post type. I guess maybe the idea is that the hook is Post editor specific yet we want the change to apply in any editor? However for the Site editor it doesn’t seem possible for the canvas to not be iframed. In the widgets editor it seems pattern insertion isn’t allowed so pattern editing is also not possible.

**Comment by t-hamano** [comment]
I've tried consolidating all the conditions into the `useShouldIframe` hook, but what do you think?

> In the widgets editor it seems pattern insertion isn’t allowed so pattern editing is also not possible.

That's a good point. The widget editor doesn't allow us to create patterns from blocks. Even if you copy and paste a sync pattern from the post editor, it doesn't do anything because `core/block` isn't registered:

image

Perhaps there are hurdles to overcome before we can add patterns to the widget editor. Related issues:

- https://github.com/WordPress/gutenberg/issues/26170
- https://github.com/WordPress/gutenberg/issues/32952

**Comment by t-hamano** [comment]
> I've tried consolidating all the conditions into the `useShouldIframe` hook

~~This appears to have caused some E2E test failures. There may be something we're missing that simply can't be integrated into `useShouldIframe()` hook.~~

~~I've reverted this commit.~~

The E2E test failure does not appear to be caused by the integration. I will try restoring the commit again at a later date.

Update: The test failure was fixed in #65939, so I committed the condition integration again.

File: packages/editor/src/components/visual-editor/index.js [hunk 1]
```diff
@@ -358,8 +358,6 @@ function VisualEditor( {
 		! isMobileViewport &&
 		// Dsiable resizing in zoomed-out mode.
 		! isZoomedOut;
-	const shouldIframe =
-		! disableIframe || [ 'Tablet', 'Mobile' ].includes( deviceType );
```
Comment (stokesman) [inline packages/editor/src/components/visual-editor/index.js:364]: When I first commented, I wasn’t understanding you meant to remove the conditions here but it does make sense to me now. Still, it looks like the tablet and mobile conditions got left out of the post editor’s `useShouldIframe` hook. I tested and the device previews weren’t working from a non-iframe starting point.

**Review by stokesman — COMMENTED** [review]
Thanks for the updates. This looks like it’s on the right track.

File: packages/editor/src/components/visual-editor/index.js (hunk 1 above)
Comment (t-hamano) [reply packages/editor/src/components/visual-editor/index.js:364]: Sorry, in my repeated commits I seem to have forgotten to update the `useShouldIframe` hook. It should work as expected now.

**Review by stokesman — APPROVED** [review]
I’ve tested again and it works as expected. There’s a conflict that popped up thanks to #66141 but it looks like it should be a straightforward resolution so I’m green lighting this anyway.

**Comment by t-hamano** [comment]
Thanks for the review! The conflict has been resolved.

**Pull request merged**